pkg/operator/target_status.go

```diff
@@ -0,0 +1,936 @@
+	// minDelay := time.NewTimer(10 * time.Second)
+
+	minDelay := r.tickerBuilder()
+	// defer minDelay.Stop()
+
+	onceLog := new(sync.Once)
+
+	// Wrapper function so we could exit early.
+	err := func() error {
+		now := time.Now()
+		if err := poll(ctx, r.logger, r.opts, r.manager); err != nil {
+			// Not having data is not an error.
+			if apierrors.IsNotFound(err) {
+				onceLog.Do(func() {
+					r.logger.Error(err, "Kubernetes config not found")
+				})
+				return nil
+			}
+			return err
+		}
+		// In case Kubernetes config is removed and re-added.
+		onceLog = new(sync.Once)
+
+		duration := time.Since(now)
+		targetStatusDuration.WithLabelValues().Set(float64(duration.Milliseconds()))
+		return nil
+	}()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if we beat the timer, otherwise wait.
+	<-minDelay.Channel()
+	// Re-trigger event
+	r.queue()
+
+	return reconcile.Result{}, nil
+}
+
+func poll(ctx context.Context, logger logr.Logger, opts Options, manager dataManager) error {
+	targets, err := fetchTargets(ctx, logger, opts, manager)
+	if err != nil {
+		return err
+	}
+
+	return populateTargets(ctx, logger, manager, targets)
+}
+
+func fetchTargets(ctx context.Context, logger logr.Logger, opts Options, manager dataManager) ([]*prometheusTargetsResultAndTime, error) {
+	podMap, err := manager.getPrometheusPods(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up pod job queue and jobs
+	podDiscoveryChan := make(chan prometheusPod, opts.TargetPollConcurrency)
+	wg := sync.WaitGroup{}
+	wg.Add(int(opts.TargetPollConcurrency))
+
+	// Must be unbounded or else we deadlock.
+	targetChan := make(chan *prometheusTargetsResultAndTime)
+	for i := uint16(0); i < opts.TargetPollConcurrency; i++ {
+		// Wrapper function to handle errors.
+		go func(thread uint16) {
+			fetcher, err := manager.getTargetFetcher(ctx, thread)
+			if err != nil {
+				logger.Error(err, "error getting target fetcher")
+				return
+			}
+			err = processTargetQueue(ctx, logger, &wg, fetcher, podDiscoveryChan, targetChan)
+			if err != nil {
+				logger.Error(err, "error processing target queue")
+			}
+		}(i)
+	}
+
+	// Unbuffered channels are blocking so make sure we end the goroutine processing them.
+	go func() {
+		for port, pods := range podMap {
+			for _, pod := range pods {
+				podDiscoveryChan <- prometheusPod{
+					port: port,
+					pod:  pod,
+				}
+			}
+		}
+
+		// Must close so jobs aren't waiting on the channel indefinitely.
+		close(podDiscoveryChan)
+
+		// Close target after we're sure all targets are queued.
+		wg.Wait()
+		close(targetChan)
+	}()
+
+	results := make([]*prometheusTargetsResultAndTime, 0)
+	for target := range targetChan {
+		results = append(results, target)
+	}
+
+	return results, nil
+}
+
+func populateTargets(ctx context.Context, logger logr.Logger, manager dataManager, targets []*prometheusTargetsResultAndTime) error {
+	endpointBuilder := newScrapeEndpointBuilder()
```

Comment:
Do we need a dedicated struct to maintain state here? In general, if we don't need to maintain state, it's easier to plumb this through a pure function. This also makes unit testing easier. i.e.
```go
statusesByJob := buildEndpointStatuses(targets)
```

Reply:
That being said, I think there is some merit in having your `scrapeEndpointStatusBuilder`. But the `buildEndpointStatuses()` function can handle the bookkeeping of setting up the `map[job][]monitoringv1.ScrapeEndpointStatuses` much like your current `scrapeEndpointBuilder` with less indirection.

I also vote moving the `scrapeEndpointStatusBuilder` to its own `endpoint_status_builder.go` file for easier parsing.